lib/puppet/provider/ec2_instance/v2.rb

```diff
@@ -46,10 +46,11 @@ def self.instance_to_hash(region, instance)
     end
     subnet_name = nil
     if instance.subnet_id
-      subnet_response = ec2_client(region).describe_subnets(subnet_ids: [instance.subnet_id])
-      subnet_name_tag = subnet_response.data.subnets.first.tags.detect { |tag| tag.key == 'Name' }
+      subnet_response = ec2_client(region).describe_subnets(filters: [
+        {name: 'subnet-id', values: [instance.subnet_id]}
+      ])
+      subnet_name = subnet_response.data.subnets.empty? ? nil : name_from_tag(subnet_response.data.subnets.first)
```

Comment:
I am familiar with the issue fixed in https://tickets.puppetlabs.com/browse/PUP-3656 / discussed on the PR at https://github.com/puppetlabs/puppet/pull/3428

I think there's a bit of confusion around `if we raise exceptions here they get eaten by Puppet`
  That's a bit imprecise, because that's only true for errors derived from `StandardError`.  We can simply create our own errors that derive directly from `Exception` and they will no longer be eaten in `prefetch`.

```ruby
class MyError < Exception
end

begin
  raise StandardError.new
rescue => foo
  puts "rescued: #{foo}"
end

begin
  raise MyError.new
rescue => bar
  puts "rescued: #{bar}"
end
```

```
[source] ruby test.rb                                                       10:27:28
rescued: StandardError
test.rb:11:in `<main>': MyError (MyError)
```

Regardless of where the state consistency issues are, I still don't believe it's a good idea to assume or otherwise act like that problem doesn't exist but papering over these details in `prefetch`